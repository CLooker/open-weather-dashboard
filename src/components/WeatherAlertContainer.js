import React, { Component } from 'react';
import CreateWeatherAlert from './CreateWeatherAlert';
import GetWeatherAlerts from './GetWeatherAlerts';

export default class WeatherAlertContainer extends Component {
  state = {
    registeredAlerts: null,
    triggeredAlerts: []
  };

  incrementRegisteredAlerts = () => {
    this.setState(({ registeredAlerts }) => ({
      registeredAlerts: ++registeredAlerts
    }));
  };

  updateRegisteredAlerts = ({ length }) => {
    this.setState({ registeredAlerts: length });
  };

  updateTriggeredAlerts = alerts => {
    this.setState(prevState => ({
      triggeredAlerts: [...prevState.triggeredAlerts, ...alerts]
    }));
  };

  render() {
    const { registeredAlerts, triggeredAlerts } = this.state;
    return (
      <div className="weather-alert-container">
        <div className="weather-alert-container-title-icon">
          <img
            src="https://cloudcheckr.com/wp-content/uploads/2015/12/alert.jpg"
            alt="alert"
          />
          <h3>Weather Alerts</h3>
        </div>
        <div className="alert-container">
          <CreateWeatherAlert
            incrementRegisteredAlerts={this.incrementRegisteredAlerts}
          />
          <GetWeatherAlerts
            registeredAlerts={registeredAlerts}
            triggeredAlerts={triggeredAlerts}
            updateRegisteredAlerts={this.updateRegisteredAlerts}
            updateTriggeredAlerts={this.updateTriggeredAlerts}
          />
        </div>
      </div>
    );
  }
}
