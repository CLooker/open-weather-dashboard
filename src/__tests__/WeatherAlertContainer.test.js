import React from 'react';
import { mount, shallow } from 'enzyme';
import WeatherAlertContainer from '../components/WeatherAlertContainer';

describe('WeatherAlertContainer', () => {
  let component;

  beforeEach(() => (component = shallow(<WeatherAlertContainer />)));

  describe('initialization', () => {
    it('should render correctly', () => {
      expect(component.debug()).toMatchSnapshot();
    });

    it('should have certain state', () => {
      const actual = component.state();
      const expected = {
        registeredAlerts: null,
        triggeredAlerts: []
      };
      expect(actual).toEqual(expected);
    });
  });

  describe('rendering', () => {
    const state = {
      registeredAlerts: '5',
      triggeredAlerts: ['hey', 'ho', "let's go"]
    };
    let mountedComponent;

    beforeEach(() => {
      mountedComponent = mount(<WeatherAlertContainer />);
      mountedComponent.setState(state);
    });

    describe('props passed to `CreateWeatherAlert`', () => {
      let actual;

      beforeEach(
        () => (actual = mountedComponent.find('CreateWeatherAlert').props())
      );

      it('should pass certain methods', () => {
        const expectedMethod = 'incrementRegisteredAlerts';
        expect(actual[expectedMethod]).toEqual(expect.any(Function));
      });
    });

    describe('props passed to `GetWeatherAlerts`', () => {
      let actual;

      beforeEach(
        () => (actual = mountedComponent.find('GetWeatherAlerts').props())
      );

      it('should pass certain props', () => {
        const { registeredAlerts, triggeredAlerts } = state;
        const expected = { registeredAlerts, triggeredAlerts };
        expect(actual).toEqual(expect.objectContaining(expected));
      });

      it('should pass certain methods', () => {
        const expected = ['updateRegisteredAlerts', 'updateTriggeredAlerts'];
        expected.forEach(method => {
          expect(actual[method]).toEqual(expect.any(Function));
        });
      });
    });
  });

  describe('methods', () => {
    describe('`updateRegisteredAlerts`', () => {
      it('shoule `setState` correctly', () => {
        const length = 5;
        component.instance().updateRegisteredAlerts({ length });
        const actual = component.state().registeredAlerts;
        const expected = length;
        expect(actual).toBe(expected);
      });
    });

    describe('`updateTriggeredAlerts`', () => {
      it('should', () => {
        const initialAlert = ['hey'];
        component.setState({ triggeredAlerts: initialAlert });
        const alerts = ['ho', "let's go"];
        component.instance().updateTriggeredAlerts(alerts);
        const actual = component.state().triggeredAlerts;
        const expected = initialAlert.concat(alerts);
        expect(actual).toEqual(expected);
      });
    });
  });
});
