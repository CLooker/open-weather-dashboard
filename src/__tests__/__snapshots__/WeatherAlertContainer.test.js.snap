// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherAlertContainer initialization should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WeatherAlertContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="weather-alert-container-title-icon"
        >
          <img
            alt="alert"
            src="alert.jpg"
          />
          <h3>
            Weather Alerts
          </h3>
        </div>,
        <div
          className="alert-container"
        >
          <CreateWeatherAlert
            incrementRegisteredAlerts={[Function]}
          />
          <GetWeatherAlerts
            registeredAlerts={null}
            triggeredAlerts={Array []}
            updateRegisteredAlerts={[Function]}
            updateTriggeredAlerts={[Function]}
          />
        </div>,
      ],
      "className": "weather-alert-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="alert"
              src="alert.jpg"
            />,
            <h3>
              Weather Alerts
            </h3>,
          ],
          "className": "weather-alert-container-title-icon",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "alert",
              "src": "alert.jpg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Weather Alerts",
            },
            "ref": null,
            "rendered": "Weather Alerts",
            "type": "h3",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <CreateWeatherAlert
              incrementRegisteredAlerts={[Function]}
            />,
            <GetWeatherAlerts
              registeredAlerts={null}
              triggeredAlerts={Array []}
              updateRegisteredAlerts={[Function]}
              updateTriggeredAlerts={[Function]}
            />,
          ],
          "className": "alert-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "incrementRegisteredAlerts": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "registeredAlerts": null,
              "triggeredAlerts": Array [],
              "updateRegisteredAlerts": [Function],
              "updateTriggeredAlerts": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="weather-alert-container-title-icon"
          >
            <img
              alt="alert"
              src="alert.jpg"
            />
            <h3>
              Weather Alerts
            </h3>
          </div>,
          <div
            className="alert-container"
          >
            <CreateWeatherAlert
              incrementRegisteredAlerts={[Function]}
            />
            <GetWeatherAlerts
              registeredAlerts={null}
              triggeredAlerts={Array []}
              updateRegisteredAlerts={[Function]}
              updateTriggeredAlerts={[Function]}
            />
          </div>,
        ],
        "className": "weather-alert-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="alert"
                src="alert.jpg"
              />,
              <h3>
                Weather Alerts
              </h3>,
            ],
            "className": "weather-alert-container-title-icon",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "alert",
                "src": "alert.jpg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Weather Alerts",
              },
              "ref": null,
              "rendered": "Weather Alerts",
              "type": "h3",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <CreateWeatherAlert
                incrementRegisteredAlerts={[Function]}
              />,
              <GetWeatherAlerts
                registeredAlerts={null}
                triggeredAlerts={Array []}
                updateRegisteredAlerts={[Function]}
                updateTriggeredAlerts={[Function]}
              />,
            ],
            "className": "alert-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "incrementRegisteredAlerts": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "registeredAlerts": null,
                "triggeredAlerts": Array [],
                "updateRegisteredAlerts": [Function],
                "updateTriggeredAlerts": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
